File: wire-android-automation/common/src/main/java/com/wearezeta/auto/common/calling2/v1/CallingServiceClient.java
package com.wearezeta.auto.common.calling2.v1;

import java.awt.image.BufferedImage;
import java.util.List;
import java.util.logging.Logger;

import com.wearezeta.auto.common.CommonUtils;
import com.wearezeta.auto.common.Config;
import com.wearezeta.auto.common.backend.BackendConnections;
import com.wearezeta.auto.common.calling2.v1.exception.CallingServiceCallException;
import com.wearezeta.auto.common.calling2.v1.exception.CallingServiceInstanceException;
import com.wearezeta.auto.common.calling2.v1.model.*;
import com.wearezeta.auto.common.log.ZetaLogger;
import com.wearezeta.auto.common.usrmgmt.ClientUser;

public class CallingServiceClient {

    private static final Logger LOG = ZetaLogger
            .getLog(CallingServiceClient.class.getSimpleName());

    private static final boolean TRACE = false;
    private final CallingService callingService = new CallingService(getApiRoot(), TRACE);

    public Instance startInstance(ClientUser userAs, String verificationCode, VersionedInstanceType instanceType,
                                  String name, boolean beta) throws CallingServiceInstanceException {
        String callingServiceEnv = Config.common().getCallingServiceEnvironment(CallingServiceClient.class);
        BackendDTO backendDTO = BackendDTO.fromBackend(BackendConnections.get(userAs));
        // set default webapp to master if test runs against master webapp, when using 'custom' calling env
        if (CommonUtils.isWebapp()) {
            if (!instanceType.isZCall() && callingServiceEnv.equalsIgnoreCase("custom")
                    && Config.current().getWebAppApplicationPath(CallingServiceClient.class).contains("master")
                    && BackendConnections.get(userAs).getBackendName().equals("staging")) {
                backendDTO = BackendDTO.fromBackend(BackendConnections.get("staging-with-webapp-master"));
            }
        }

        InstanceRequest instanceRequest = new InstanceRequest(
                userAs.getEmail(), userAs.getPassword(),
                verificationCode,
                callingServiceEnv,
                backendDTO,
                instanceType, name, beta,  1000L * 60 * 10);
        return callingService.createInstance(instanceRequest);
    }

    public Instance stopInstance(Instance instance)
            throws CallingServiceInstanceException {
        return callingService.destroyInstance(instance);
    }

    public InstanceStatus getInstanceStatus(Instance instance)
            throws CallingServiceInstanceException {
        return callingService.getInstance(instance).getStatus();
    }

    public Call acceptNextIncomingCall(Instance instance)
            throws CallingServiceCallException {
        CallRequest callRequest = new CallRequest();

        return callingService.acceptNext(instance, callRequest);
    }

    public Call acceptNextIncomingVideoCall(Instance instance)
            throws CallingServiceCallException {
        CallRequest callRequest = new CallRequest();

        return callingService.acceptNextVideo(instance, callRequest);
    }

    public Call callToUser(Instance instance, String convId)
            throws CallingServiceCallException {
        CallRequest callRequest = new CallRequest(convId);
        return callingService.start(instance, callRequest);
    }

    public Call videoCallToUser(Instance instance, String convId)
            throws CallingServiceCallException {
        CallRequest callRequest = new CallRequest(convId);
        return callingService.startVideo(instance, callRequest);
    }

    public Call getCall(Instance instance, Call call)
            throws CallingServiceCallException {
        return callingService.getCall(instance, call);
    }

    public Call stopCall(Instance instance, Call call)
            throws CallingServiceCallException {
        return callingService.stop(instance, call);
    }
    
    public Call declineCall(Instance instance, String convId)
            throws CallingServiceCallException {
        CallRequest callRequest = new CallRequest(convId);
        return callingService.decline(instance, callRequest);
    }

    // TODO: mute/unmute/listen/speak

    private static String getApiRoot() {
        return Config.common().getCallingServiceUrl(CallingServiceClient.class);
    }

    public List<Flow> getFlows(Instance instance)
            throws CallingServiceInstanceException {
        return callingService.getFlows(instance);
    }

    public String getPackets(Instance instance)
            throws CallingServiceInstanceException {
        return callingService.getPackets(instance);
    }

    public String getLog(Instance instance)
            throws CallingServiceInstanceException {
        return callingService.getLog(instance);
    }

    public Call switchVideoOn(Instance instance, Call call)
            throws CallingServiceCallException {
        return callingService.switchVideoOn(instance, call);
    }

    public Call switchVideoOff(Instance instance, Call call)
            throws CallingServiceCallException {
        return callingService.switchVideoOff(instance, call);
    }

    public Call pauseVideoCall(Instance instance, Call call)
            throws CallingServiceCallException {
        return callingService.pauseVideoCall(instance, call);
    }

    public Call unpauseVideoCall(Instance instance, Call call)
            throws CallingServiceCallException {
        return callingService.unpauseVideoCall(instance, call);
    }

    public Call switchScreensharingOn(Instance instance, Call call)
            throws CallingServiceCallException {
        return callingService.switchScreensharingOn(instance, call);
    }

    public Call switchScreensharingOff(Instance instance, Call call)
            throws CallingServiceCallException {
        return callingService.switchScreensharingOff(instance, call);
    }

    public Call getCurrentCall(Instance instance) throws CallingServiceInstanceException {
        return callingService.getInstance(instance).getCurrentCall();
    }

    public String getLivePreview(Instance instance) throws CallingServiceInstanceException {
        return callingService.getInstance(instance).getLivePreview();
    }

    public BufferedImage getScreenshot(Instance instance) throws CallingServiceInstanceException {
        return callingService.getScreenshot(instance);
    }

    public Call muteMicrophone(Instance instance, Call call)
            throws CallingServiceCallException {
        return callingService.muteMicrophone(instance, call);
    }

    public Call unmuteMicrophone(Instance instance, Call call)
            throws CallingServiceCallException {
        return callingService.unmuteMicrophone(instance, call);
    }

    public Call maximiseCall(Instance instance, Call call)
            throws CallingServiceCallException {
        return callingService.maximiseCall(instance, call);
    }

    public List<Instance> getAllRunningInstances() {
        return callingService.getAllRunningInstances();
    }

    // region moderation

    public Call muteParticipant(Instance instance, Call call, String name) throws CallingServiceCallException {
        return callingService.muteParticipantX(instance, call, name);
    }

    public Call muteAllOthers(Instance instance, Call call, String name) throws CallingServiceCallException {
        return callingService.muteAllOthers(instance, call, name);
    }

    // endregion
}

File: tests/testsSupport/src/main/call/CallingServiceClient.kt
/*
 * Wire
 * Copyright (C) 2025 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses/.
 */
@file:Suppress(
    "TooGenericExceptionCaught",
    "LargeClass",
    "LongParameterList",
    "NestedBlockDepth",
    "MagicNumber",
    "TooManyFunctions",
    "TooGenericExceptionThrown"
)
package call

import android.graphics.Bitmap
import backendUtils.BackendClient
import call.models.BackendDTO
import call.models.Call
import call.models.CallRequest
import call.models.CallFlow
import call.models.Instance
import call.models.InstanceRequest
import call.models.InstanceStatus
import call.models.VersionedInstanceType
import user.utils.ClientUser

class CallingServiceClient {

    companion object {
        const val API_ROOT = "https://qa-callingservice-wire.runs.onstackit.cloud"
    }

    private val callingService = CallingService(API_ROOT)

    fun startInstance(
        user: ClientUser,
        verificationCode: String,
        instanceType: VersionedInstanceType,
        name: String,
        beta: Boolean
    ): Instance {
        val callingServiceEnv = "custom"

        var backendDTO = BackendDTO.fromBackend(BackendClient.loadBackend(user.backendName.orEmpty()))

        val request = InstanceRequest(
            email = user.email.orEmpty(),
            password = user.password.orEmpty(),
            verificationCode = verificationCode,
            backend = callingServiceEnv,
            customBackend = backendDTO,
            instanceType = instanceType,
            name = name,
            beta = beta,
            timeout = 1000L * 60 * 10
        )

        return callingService.createInstance(request)
    }

    fun stopInstance(instance: Instance): Instance =
        callingService.destroyInstance(instance)

    fun getInstanceStatus(instance: Instance): InstanceStatus? =
        callingService.getInstance(instance).instanceStatus

    fun acceptNextIncomingCall(instance: Instance): Call =
        callingService.acceptNext(instance, CallRequest())

    fun acceptNextIncomingVideoCall(instance: Instance): Call =
        callingService.acceptNextVideo(instance, CallRequest())

    fun callToUser(instance: Instance, convId: String): Call =
        callingService.start(instance, CallRequest(convId))

    fun videoCallToUser(instance: Instance, convId: String): Call =
        callingService.startVideo(instance, CallRequest(convId))

    fun getCall(instance: Instance, call: Call): Call =
        callingService.getCall(instance, call)

    fun stopCall(instance: Instance, call: Call): Call =
        callingService.stop(instance, call)

    fun declineCall(instance: Instance, convId: String): Call =
        callingService.decline(instance, CallRequest(convId))

    fun getFlows(instance: Instance): List<CallFlow> =
        callingService.getFlows(instance)

    fun getPackets(instance: Instance): String =
        callingService.getPackets(instance)

    fun getLog(instance: Instance): String =
        callingService.getLog(instance)

    fun switchVideoOn(instance: Instance, call: Call): Call =
        callingService.switchVideoOn(instance, call)

    fun switchVideoOff(instance: Instance, call: Call): Call =
        callingService.switchVideoOff(instance, call)

    fun pauseVideoCall(instance: Instance, call: Call): Call =
        callingService.pauseVideoCall(instance, call)

    fun unpauseVideoCall(instance: Instance, call: Call): Call =
        callingService.unpauseVideoCall(instance, call)

    fun switchScreensharingOn(instance: Instance, call: Call): Call =
        callingService.switchScreensharingOn(instance, call)

    fun switchScreensharingOff(instance: Instance, call: Call): Call =
        callingService.switchScreensharingOff(instance, call)

    fun getCurrentCall(instance: Instance): Call? =
        callingService.getInstance(instance).currentCall

    fun getLivePreview(instance: Instance): String? =
        callingService.getInstance(instance).screenshot

    fun getScreenshot(instance: Instance): Bitmap? =
        callingService.getScreenshot(instance)

    fun muteMicrophone(instance: Instance, call: Call): Call =
        callingService.muteMicrophone(instance, call)

    fun unmuteMicrophone(instance: Instance, call: Call): Call =
        callingService.unmuteMicrophone(instance, call)

    fun maximiseCall(instance: Instance, call: Call): Call =
        callingService.maximiseCall(instance, call)

    fun getAllRunningInstances(): List<Instance> =
        callingService.getAllRunningInstances()

    // region moderation

    fun muteParticipant(instance: Instance, call: Call, name: String): Call =
        callingService.muteParticipant(instance, call, name)

    fun muteAllOthers(instance: Instance, call: Call, name: String): Call =
        callingService.muteAllOthers(instance, call, name)

    // endregion
}
